recipes/binutils/all/conanfile.py

```diff
@@ -126,64 +129,78 @@ def _raise_unsupported_configuration(self, key, value):
+    def generate(self):
         yes_no = lambda tf : "yes" if tf else "no"
-        conf_args = [
-            f"--target={self.options.target_triplet}",
-            f"--enable-multilib={yes_no(self.options.multilib)}",
-            "--with-system-zlib",
-            "--disable-nls",
-            f"--program-prefix={self.options.prefix}",
-            f"exec_prefix={tools.unix_path(self._exec_prefix)}",
-        ]
-        autotools.configure(args=conf_args, configure_dir=self._source_subfolder)
-        return autotools
+        tc = AutotoolsToolchain(self)
+        tc.configure_args.append("--disable-nls")
+        tc.configure_args.append(f"--target={self.options.target_triplet}")
+        tc.configure_args.append(f"--enable-multilib={yes_no(self.options.multilib)}")
+        tc.configure_args.append(f"--with-zlib={self.deps_cpp_info['zlib'].rootpath}")
+        tc.configure_args.append(f"--program-prefix={self.options.prefix}")
+        tc.configure_args.append(f"--exec_prefix={unix_path(self, os.path.join('/','bin', 'exec_prefix'))}")
+        tc.generate()
 
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
-        autotools = self._configure_autotools()
+        for _patch in self.conan_data.get("patches", {}).get(self.version, []):
+            patch(self, **_patch)
+        autotools = Autotools(self)
+        autotools.configure()
         autotools.make()
 
     def package(self):
-        self.copy("COPYING*", src=self._source_subfolder, dst="licenses")
-        autotools = self._configure_autotools()
+        autotools = Autotools(self)
         autotools.install()
 
-        tools.rmdir(os.path.join(self.package_folder, "share"))
-
-        tools.remove_files_by_mask(os.path.join(self.package_folder, "lib"), "*.la")
+        rmdir(self, os.path.join(self.package_folder, "share"))
+        rm(self, "*.la", os.path.join(self.package_folder, "lib"), recursive=True)
+        copy(
+            self,
+            pattern="COPYING*",
+            dst=os.path.join(self.package_folder, "licenses"),
+            src=self.source_folder,
+            keep_path=False,
+        )
 
     def package_info(self):
+        target_bindir = os.path.join(self._exec_prefix, str(self.options.target_triplet), "bin")
+        self.output.info("Prepending PATH environment variable: {}".format(target_bindir))
+        self.buildenv_info.prepend_path("PATH", target_bindir)
+
+        binaries = os.listdir(target_bindir)
+        self.output.info(f"Binaries built: {', '.join(binaries)}")
+        for binary_name in binaries:
+            binary = os.path.join(target_bindir, binary_name)
+            if os.path.isfile(binary):
+                self.output.info(f"Setting {binary_name.upper()}={binary}")
+                self.buildenv_info.define(f"{binary_name.upper()}", binary)
```

Comment:
What is this exporting exacly?

IT's better to pt a more explicit list - this way we can review and understand the implaations. 🙏

Reply:
the binaries include `["ar", "nm", "objcopy", "objdump", "ranlib", "readelf", "strip", "as", "ld"]` on linux. Which means that this just exports those names in upper case, i.e:
* `AR={package_folder}/bin/exec_prefix/x86_64-pc-linux-gnu/bin/ar`
* `NM={package_folder}/bin/exec_prefix/x86_64-pc-linux-gnu/bin/nm`
* `OBJCOPY={package_folder}/bin/exec_prefix/x86_64-pc-linux-gnu/bin/objcopy`

The reason I chose to resolve this programmatically is because the collection of binaries built is reasonably dependent on the platform and architecture it is building for. To see the logic that it uses to choose what targets are built, see:

https://gitlab.com/git-mirrors/binutils/-/blob/master/configure#L3804-4103

To illustrate with an example, on x86_64 macOS, `binutils` will build everything in the above list with the exception of the assembler (`as`) and linker (`ld`) because these aren't supported on that platform. Anyway, the logic here simplifies the export logic reasonably substantially by exporting all binaries that were built with a variable name to match the uppercase name of the target.

I think that this is pretty readable - if `binary_name.upper()` isn't self documenting, I'm not sure what is. I think that this is the best way to deal with this though, I can add a comment to clarify the binaries that would commonly be included and an example of the environment variable if you'd like? These are also printed to the console, so it would be reasonably clear to the user.

Reply:
I've landed here from the comment left in the recipe :) 

I think that reading the recipe alone it's not entirely clear (as a reader), what this is doing fully, in order to work this out, one must remember which executables are part of the binutils package, and there's several.

The secondary thing is that as a reviewer, it's also not clear _why_ these variables need to be defined and why they need to be absolute paths - can't downstream consumers not rely on `PATH` being defined during `./configure` to find the correct tools inside _this_ package?

I think don't think there are problems with readability per se, but perhaps if there had been a explicit list, we could have collectively identified this https://github.com/conan-io/conan-center-index/issues/14566 before it happened.